catalogs/catalog-lakehouse-iceberg/src/main/java/com/datastrato/gravitino/catalog/lakehouse/iceberg/backend/HiveBackendProxy.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.hadoop.hive.thrift.DelegationTokenIdentifier;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.Token;
+import org.apache.iceberg.ClientPool;
+import org.apache.iceberg.hive.HiveCatalog;
+import org.apache.thrift.TException;
+
+/**
+ * Proxy class for HiveCatalog to support kerberos authentication. We can also make HiveCatalog as a
+ * generic type and pass it as a parameter to the constructor.
+ */
+public class HiveBackendProxy implements MethodInterceptor {
+
+  private final HiveCatalog target;
+  private String kerberosRealm;
+  private final UserGroupInformation proxyUser;
+  private final Map<String, String> properties;
+
+  public HiveBackendProxy(Map<String, String> properties, HiveCatalog target) {
+    this.target = target;
+    this.properties = properties;
+    try {
+      initKerberos(properties, target.getConf());
+      proxyUser = UserGroupInformation.getCurrentUser();
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to get current user", e);
+    }
+  }
+
+  private void initKerberos(Map<String, String> properties, Configuration conf) {
+    try {
+      KerberosClient kerberosClient = new KerberosClient(properties, conf);
+      File keytabFile =
+          kerberosClient.saveKeyTabFileFromUri(Long.valueOf(properties.get("catalog_uuid")));
+      this.kerberosRealm = kerberosClient.login(keytabFile.getAbsolutePath());
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to login with kerberos", e);
+    }
+  }
+
+  @Override
+  public Object intercept(Object o, Method method, Object[] objects, MethodProxy methodProxy)
+      throws Throwable {
+
+    String proxyKerberosPrincipalName = PrincipalUtils.getCurrentPrincipal().getName();
+    if (!proxyKerberosPrincipalName.contains("@")) {
+      proxyKerberosPrincipalName =
+          String.format("%s@%s", proxyKerberosPrincipalName, kerberosRealm);
+    }
+
+    UserGroupInformation realUser =
+        UserGroupInformation.createProxyUser(proxyKerberosPrincipalName, proxyUser);
+
+    String token = getToken();
+
+    Token<DelegationTokenIdentifier> delegationToken = new Token<>();
+    delegationToken.decodeFromUrlString(token);
+    realUser.addToken(delegationToken);
+
+    return realUser.doAs(
+        (PrivilegedExceptionAction<Object>)
+            () -> {
+              try {
+                return methodProxy.invoke(target, objects);
+              } catch (Throwable e) {
+                if (RuntimeException.class.isAssignableFrom(e.getClass())) {
+                  throw (RuntimeException) e;
+                }
+                throw new RuntimeException("Failed to invoke method", e);
+              }
+            });
+  }
+
+  private String getToken()
+      throws NoSuchFieldException, IllegalAccessException, TException, InterruptedException {
+    final Field m = HiveCatalog.class.getDeclaredField("clients");
```

Comment:
Is Iceberg class isolated? Could we put these in the initialization to avoid thread issues?

Reply:
> Is Iceberg class isolated?

What do you mean? Can you clarify it more clearly?

Reply:
HiveCatalog clients is a static field. Will it be influenced by different catalogs?

Reply:
![image](https://github.com/datastrato/gravitino/assets/15794564/81df572c-8304-4e89-b04e-9b96c686c5e0)